core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -511,6 +514,28 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
       uncleanablePartitions.get(log.parentDir).exists(partitions => partitions.contains(topicPartition))
     }
   }
+
+  def maintainUncleanablePartitions(): Unit = {
+    // Remove deleted partitions from uncleanablePartitions
+    inLock(lock) {
+      // Remove non-existing logDir
+      // Note: we don't use retain or filterInPlace method in this function because retain in deprecated in
+      // scala 2.13 while filterInPlace is not available in scala 2.12.
+      uncleanablePartitions.filterNot {
+        case (logDir, _) => logDirs.map(_.getAbsolutePath).contains(logDir)
+      }.keys.foreach {
+        uncleanablePartitions.remove(_)
+      }
+
+      uncleanablePartitions.values.foreach {
+        // Remove deleted partitions
+        partitions => partitions.filterNot(logs.contains(_)).foreach {
+          partitions.remove(_)
```

Comment:
Hmm, does removing from a set has side effect of an ongoing iterator? If so, we want to figure out the items to remove first and do the removal in the end. Ditto for the code above in line 527.

Reply:
Good catch. I thought "partitions.filterNot(logs.contains(_))" will generate a temporary collection first and then "foreach" iterates over the temporary collection.   It's better to remove this implicit assumption.